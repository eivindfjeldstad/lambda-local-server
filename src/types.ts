import { AuthResponseContext } from 'aws-lambda';

export type OptionalApiGatewayEventRequestContext = {
  accountId?: string;
  apiId?: string;
  authorizer?: AuthResponseContext | null;
  identity?: {
    accessKey: string | null;
    accountId: string | null;
    apiKey: string | null;
    apiKeyId: string | null;
    caller: string | null;
    cognitoAuthenticationProvider: string | null;
    cognitoAuthenticationType: string | null;
    cognitoIdentityId: string | null;
    cognitoIdentityPoolId: string | null;
    sourceIp: string;
    user: string | null;
    userAgent: string | null;
    userArn: string | null;
  };
  stage?: string;
  requestId?: string;
  resourceId?: string;
  resourcePath?: string;
};

export type CognitoIdResolver =
  | (() => string | Promise<string>)
  | string
  | undefined;

export type LambdaOptions = {
  entry: string;
  contextPath?: string;
  urls?: string[];
  mockHeaders?: { [key: string]: string };
  context?: OptionalApiGatewayEventRequestContext;
  cognitoId?: CognitoIdResolver;
};

export type AppOptions = {
  lambdas: LambdaOptions[];
  path?: string;
  context?: OptionalApiGatewayEventRequestContext;
  cognitoId?: CognitoIdResolver;
  cacheNodeModules?: boolean;
  onCacheCleared?: () => void;
};

export type LambdaServer = {
  listen: (port?: number) => Promise<number>;
  close: () => void;
};
