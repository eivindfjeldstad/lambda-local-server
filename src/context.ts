import { APIGatewayEventRequestContext } from 'aws-lambda';
import { Request } from 'express';
import { LambdaOptions, AppOptions, CognitoIdResolver } from './types';

const contextDefault = {
  accountId: 'test',
  apiId: 'test',
  authorizer: undefined,
  identity: undefined,
  stage: 'test',
  requestId: 'test',
  resourceId: 'test',
  resourcePath: 'test',
};

async function resolveCognitoId(resolver: CognitoIdResolver) {
  if (typeof resolver === 'string') {
    return resolver;
  } else if (typeof resolver === 'function') {
    return resolver();
  }
  return undefined;
}

async function createIdentity(
  appCognitoId: CognitoIdResolver,
  lambdaCognitoId: CognitoIdResolver
) {
  let cognitoId;
  if (lambdaCognitoId) {
    cognitoId = await resolveCognitoId(lambdaCognitoId);
  } else if (appCognitoId) {
    cognitoId = await resolveCognitoId(appCognitoId);
  }

  return {
    accessKey: 'test',
    accountId: 'test',
    apiKey: 'test',
    apiKeyId: 'test',
    caller: 'test',
    cognitoAuthenticationProvider: `test:${cognitoId}`,
    cognitoAuthenticationType: 'test',
    cognitoIdentityId: 'test',
    cognitoIdentityPoolId: 'test',
    sourceIp: 'test',
    user: 'test',
    userAgent: 'test',
    userArn: 'test',
  };
}

export async function createContext(
  appOptions: AppOptions,
  lambdaOptions: LambdaOptions,
  req: Request
): Promise<APIGatewayEventRequestContext> {
  const appContext = appOptions.context || {};
  const lambdaContext = lambdaOptions.context || {};
  return {
    ...contextDefault,
    ...appContext,
    ...lambdaContext,
    path: req.path,
    requestTimeEpoch: Date.now(),
    httpMethod: req.method,
    identity: {
      ...appContext.identity,
      ...lambdaContext.identity,
      ...(await createIdentity(appOptions.cognitoId, lambdaOptions.cognitoId)),
    },
  };
}
