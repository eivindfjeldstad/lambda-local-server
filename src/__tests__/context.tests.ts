import { Request } from 'express';

import { createContext } from '../context';
import { AppOptions, LambdaOptions } from '../types';

const DATE = new Date(2012, 12, 12).getTime();
beforeAll(() => {
  global.Date.now = jest.fn(() => DATE);
});
afterAll(() => {
  (global.Date.now as jest.Mock).mockRestore();
});

test('createContext should create context', async () => {
  const lambda: LambdaOptions = {
    entry: 'test',
  };
  const app: AppOptions = {
    lambdas: [lambda],
  };
  // @ts-ignore
  const req: Request = { path: '/', method: 'POST' };

  expect(await createContext(app, lambda, req)).toMatchSnapshot();
});

test('createContext should set cognito id from app options', async () => {
  const lambda: LambdaOptions = {
    entry: 'test',
  };
  const app: AppOptions = {
    lambdas: [lambda],
    cognitoId: 'cognito id from app',
  };
  // @ts-ignore
  const req: Request = { path: '/', method: 'POST' };

  const context = await createContext(app, lambda, req);

  expect(context.identity.cognitoAuthenticationProvider).toEqual(
    'test:cognito id from app'
  );
});

test('createContext should set cognito id from lambda options', async () => {
  const lambda: LambdaOptions = {
    entry: 'test',
    cognitoId: 'cognito id from lambda',
  };
  const app: AppOptions = {
    lambdas: [lambda],
    cognitoId: 'cognito id from app',
  };
  // @ts-ignore
  const req: Request = { path: '/', method: 'POST' };

  const context = await createContext(app, lambda, req);

  expect(context.identity.cognitoAuthenticationProvider).toEqual(
    'test:cognito id from lambda'
  );
});

test('createContext should set cognito id from function in app options', async () => {
  const lambda: LambdaOptions = {
    entry: 'test',
  };
  const app: AppOptions = {
    lambdas: [lambda],
    cognitoId: () => 'cognito id from app',
  };
  // @ts-ignore
  const req: Request = { path: '/', method: 'POST' };

  const context = await createContext(app, lambda, req);

  expect(context.identity.cognitoAuthenticationProvider).toEqual(
    'test:cognito id from app'
  );
});

test('createContext should set cognito id from function in lambda options', async () => {
  const lambda: LambdaOptions = {
    entry: 'test',
    cognitoId: () => 'cognito id from lambda',
  };
  const app: AppOptions = {
    lambdas: [lambda],
    cognitoId: () => 'cognito id from app',
  };
  // @ts-ignore
  const req: Request = { path: '/', method: 'POST' };

  const context = await createContext(app, lambda, req);

  expect(context.identity.cognitoAuthenticationProvider).toEqual(
    'test:cognito id from lambda'
  );
});
