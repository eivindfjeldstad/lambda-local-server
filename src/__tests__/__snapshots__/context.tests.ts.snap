// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContext should create context 1`] = `
Object {
  "accountId": "test",
  "apiId": "test",
  "authorizer": undefined,
  "httpMethod": "POST",
  "identity": Object {
    "accessKey": "test",
    "accountId": "test",
    "apiKey": "test",
    "apiKeyId": "test",
    "caller": "test",
    "cognitoAuthenticationProvider": "test:undefined",
    "cognitoAuthenticationType": "test",
    "cognitoIdentityId": "test",
    "cognitoIdentityPoolId": "test",
    "sourceIp": "test",
    "user": "test",
    "userAgent": "test",
    "userArn": "test",
  },
  "path": "/",
  "requestId": "test",
  "requestTimeEpoch": 1357945200000,
  "resourceId": "test",
  "resourcePath": "test",
  "stage": "test",
}
`;
